clippy_lints/src/needless_as_bytes.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::sugg::Sugg;
+use clippy_utils::ty::is_type_lang_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// It detects useless calls to `str::as_bytes()` before calling `len()` or `is_empty()`.
+    ///
+    /// ### Why is this bad?
+    /// The `len()` and `is_empty()` methods are also directly available on strings, and they
+    /// return identical results. In particular, `len()` on a string returns the number of
+    /// bytes.
+    ///
+    /// ### Example
+    /// ```
+    /// let len = "some string".as_bytes().len();
+    /// let b = "some string".as_bytes().is_empty();
+    /// ```
+    /// Use instead:
+    /// ```
+    /// let len = "some string".len();
+    /// let b = "some string".is_empty();
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub NEEDLESS_AS_BYTES,
+    complexity,
+    "detect useless calls to `as_bytes()`"
+}
+
+declare_lint_pass!(NeedlessAsBytes => [NEEDLESS_AS_BYTES]);
+
+impl LateLintPass<'_> for NeedlessAsBytes {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let ExprKind::MethodCall(method2, receiver2, &[], _) = expr.kind
+            && cx.typeck_results().expr_ty_adjusted(receiver2).peel_refs().is_slice()
+            && (method2.ident.name == sym::len || method2.ident.name.as_str() == "is_empty")
+            && let ExprKind::MethodCall(method1, receiver1, &[], _) = receiver2.kind
+            && let ty1 = cx.typeck_results().expr_ty_adjusted(receiver1).peel_refs()
+            && (is_type_lang_item(cx, ty1, LangItem::String) || ty1.is_str())
+            && method1.ident.name.as_str() == "as_bytes"
```

Comment:
You will need some extra checks here because users can implement their own trait containing an `as_bytes` method inside it and using it. The check would look like this:

```
let Some(func_id) = cx.typeck_results().type_dependent_def_id(func.hir_id)
 && let Some(impl_id) = cx.tcx.impl_of_method(func_id)
 && let None = cx.tcx.impl_trait_ref(impl_id) // An inherent impl
 && cx.tcx.type_of(impl_id).instantiate_identity()
// check that it's either str or String here
```

Also please add a test checking that case. :)

Reply:
@GuillaumeGomez I'm not sure I understand what you're saying. It looks like the tests above already catches the following case (which isn't linted):

```rust
#![warn(clippy::needless_as_bytes)]

fn main() {
    // No lint here as `.as_bytes()` doesn't come from `str` or `String`
    let _ = (&String::new() as &dyn AsBytes).as_bytes().len();
}

pub trait AsBytes {
    fn as_bytes(&self) -> &[u8];
}

impl AsBytes for String {
    fn as_bytes(&self) -> &[u8] {
        &[]
    }
}
```

Or did I miss the case you had in mind?

Reply:
No I think you're right. Missed it, sorry about that.

However, please add the code you wrote in the ui test.

Reply:
It was an interesting exercise nevertheless. I've added the extra tests.